### PR TITLE
Update test script

### PR DESCRIPTION
**Pull request by pocke**
This PR updates the test scripts system to avoid arbitrary code execution on testing. It is not a security risk for now, but this change is necessary for the new PR merging system I'm constructing.

This PR also includes the following improvements.

* Avoid duplication of `_scripts/test`
  * Previously, all `_scripts/test` files had the almost same content.
* Avoid listing dependencies by using the local source of `rbs collection`.

# How

Preivously, `_scripts/test` files in each gem had testing logic. This PR removed the script and moved the logic to `bin/test` script.

To avoid code execution, Ruby DSL files (`Steepfile` and `Gemfile`) are generated by `bin/test` script.

I also add `metadata.yaml` file to declare test dependencies. If the gem has dependencies only for the testing, the user needs to write the dependencies to `_test/matadata.yaml` file.
This file is only for testing, so it does not affect the RBS user.

### CODE REVIEW

File: bin/setup
Comment (pocke) [inline bin/setup:9]: We must install `rbs` gem globally to execute RBS without bundler context.
It is necessary for here: https://github.com/ruby/gem_rbs_collection/pull/502/files#diff-4e821476d5e4f5626c6f9fca337a47afe96a5e3daf23bccb02ad0bb109a9a407R106


`rbs collection install` command needs `Gemfile.lock` that includes the test target gems to generate `rbs_collection.lock.yaml`. But the `rbs` command should be executed in the different Bundler context than the `Gemfile.lock`. Because the `Gemfile.lock` for `rbs_collection.lock.yaml` cannot contain `rbs` gem. This `Gemfile.lock` is just generated by `bundle lock` command, so no gems are installed from the `Gemfile.lock`.





I think it is better if we can avoid the global installation. But I have no idea to avoid it 😓

Comment (soutaro) [reply bin/setup:9]: You may want to give [`bundler-compose`](https://github.com/rubygems/bundler-compose) a try?

**Review by soutaro — APPROVED** [review]
👍

**Review by ksss — APPROVED** [review]
Looks good to me.
But, I feel that it would be good to provide guidelines on how to respond to existing PR authors.

File: bin/setup
Comment (pocke) [reply bin/setup:9]: We've discussed `bundler-compose` and it is not appropriate for this situation.

* We require gems must not be installed, but `bundler-compose` wants to install gems
* If we use `bundler-compose`, we need to install `bundler-compose` gem globally

Another point, `rbs` gem is always installed globally because it is a bundled gem. So, the global installation does not have a large impact.

BTW, if the user installs Ruby with an OS package manager, `gem install rbs` will require `sudo` permission. This command will not work well in this case, but I think it is a rare case.

File: bin/test
Comment (ybiquitous) [inline bin/test:9]:
```suggestion

examples = <<~MSG
Examples:
    Test a gem
    $ bin/test gems/rack/2.2

    Test a gem with verbose messages
    $ bin/test gems/rack/2.2 -v

    Test all the gems
    $ bin/test -a
MSG
```

File: bin/test
Comment (ybiquitous) [inline bin/test:14]:
```suggestion
  opts.on("-v", "--verbose", "Print more messages") do @verbose = true end

  opts.on("-h", "--help", "Prints help") do
    puts "#{opts.help}\n#{examples}"
    exit
  end
```

File: bin/test
Comment (ybiquitous) [inline bin/test:15]:
```suggestion
end.tap do |opts|
  opts.parse!(ARGV)
end.tap do |opts|
  abort "#{opts.help}\n#{examples}" unless !ARGV.empty? || @all
end
```

**Review by ybiquitous — COMMENTED** [review]
Thanks for the pull request. Great improvements!

I have a suggestion for `bin/test`: how about putting help message out to help newcomers?

E.g.

```sh-session
$ bin/test
Usage: bin/test [options] <gem...>
    -a, --all                        Test all gems
    -v, --verbose                    Print more messages
    -h, --help                       Prints help

Examples:
    Test a gem
    $ bin/test gems/rack/2.2

    Test a gem with verbose messages
    $ bin/test gems/rack/2.2 -v

    Test all the gems
    $ bin/test -a
```

**Comment by pocke** [comment]
> But, I feel that it would be good to provide guidelines on how to respond to existing PR authors.


@ksss 


I'll take over existing PRs. I think I can merge most PRs. If I cannot merge a PR, I'll comment how to treat this change into the individual PR. 


> I have a suggestion for bin/test: how about putting help message out to help newcomers?



@ybiquitous 

Thanks for your suggestion. It looks good!
But I'd like to apply this change in another PR after this PR merges. The CI process for this PR is really heavy because the CI runs for all gems on this PR.
I'll create a PR with your suggestion. Thanks.

**Comment by ybiquitous** [comment]
> I'll create a PR with your suggestion. Thanks.

Never mind. Glad to hear! 👍🏼

File: docs/CONTRIBUTING.md
Comment (ybiquitous) [inline docs/CONTRIBUTING.md:37]:
```suggestion
We recommend adding `_test` and `_src` directories.
```

File: docs/CONTRIBUTING.md
Comment (ybiquitous) [inline docs/CONTRIBUTING.md:85]:
```suggestion
1. Write a Ruby program which uses the gem code.
```

File: docs/CONTRIBUTING.md
Comment (ybiquitous) [inline docs/CONTRIBUTING.md:87]:
```suggestion
2. Execute `bin/test gems/GEM_NAME/VERSION` to run the test suites, including `rbs validate` and `steep check`.
```

**Review by ybiquitous — APPROVED** [review]
I have another suggestion for the contributing guide; it includes typo fixes and an improvement for wording.

But of course, we could accept it in the following PR if this suggestion would be received.

Great job! LGTM 👍🏼

**Review by yykamei — APPROVED** [review]
This change looks great! I'm happy that we don't have to manage new dependencies in `_scripts/test` with  something like `-rrack`.

**Pull request merged**
